pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -255,6 +281,124 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+// The Revision of the returned ControllerRevision is set to revision. If the returned error is nil, the returned
+// ControllerRevision is valid. StatefulSet revisions are stored as patches that re-apply the current state of set
+// to a new StatefulSet using a strategic merge patch to replace the saved state of the new StatefulSet.
+func newRevision(set *apps.StatefulSet, revision int64) (*apps.ControllerRevision, error) {
+	patch, err := getPatch(set)
+	if err != nil {
+		return nil, err
+	}
+	selector, err := metav1.LabelSelectorAsSelector(set.Spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+	return history.NewControllerRevision(set,
+		controllerKind,
+		selector,
+		runtime.RawExtension{Raw: patch},
+		revision)
+}
+
+// applyRevision returns a new StatefulSet constructed by restoring the state in revision to set. If the returned error
+// is nil, the returned StatefulSet is valid.
+func applyRevision(set *apps.StatefulSet, revision *apps.ControllerRevision) (*apps.StatefulSet, error) {
+	obj, err := scheme.Scheme.DeepCopy(set)
+	if err != nil {
+		return nil, err
+	}
+	clone := obj.(*apps.StatefulSet)
+	patched, err := strategicpatch.StrategicMergePatch([]byte(runtime.EncodeOrDie(patchCodec, clone)), revision.Data.Raw, clone)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(patched, clone)
+	if err != nil {
+		return nil, err
+	}
+	return clone, nil
+}
+
+// nextRevision finds the next valid revision number based on revisions. If the length of revisions
+// is 0 this is 1. Otherwise, it is 1 greater than the largest revision's Revision. This method
+// assumes that revisions has been sorted by Revision.
+func nextRevision(revisions []*apps.ControllerRevision) int64 {
+	count := len(revisions)
+	if count <= 0 {
+		return 1
+	}
+	return revisions[count-1].Revision + 1
+}
+
+// inconsistentStatus returns true if the ObservedGeneration of status is greater than set's
+// Generation or if any of the status's fields do not match those of set's status.
+func inconsistentStatus(set *apps.StatefulSet, status *apps.StatefulSetStatus) bool {
+	return set.Status.ObservedGeneration == nil ||
+		*status.ObservedGeneration > *set.Status.ObservedGeneration ||
+		status.Replicas != set.Status.Replicas ||
+		status.CurrentReplicas != set.Status.CurrentReplicas ||
+		status.ReadyReplicas != set.Status.ReadyReplicas ||
+		status.UpdatedReplicas != set.Status.UpdatedReplicas ||
+		status.CurrentRevision != set.Status.CurrentRevision ||
+		status.UpdateRevision != set.Status.UpdateRevision
+}
+
+// completeRollingUpdate completes a rolling update when all of set's replica Pods have been updated
+// to the updateRevision. status's currentRevision is set to updateRevision and its' updateRevision
+// is set to the empty string. status's currentReplicas is set to updateReplicas and its updateReplicas
+// are set to 0.
+func completeRollingUpdate(set *apps.StatefulSet, status *apps.StatefulSetStatus) {
```

Comment:
@krmayankk you should be able to estimate if a partial rollout is complete with 
```
foo.spec.replicas-foo.spec.updateStrategy.partition.ordinal == foo.status.updatedReplicas &&
foo.spec.replicas == foo.status.readyReplicas
```
There is definitely value in communicating that via a Condition in the future.